onnxruntime/core/framework/execution_frame.cc

```diff
@@ -123,15 +121,36 @@ void IExecutionFrame::Init(const std::vector<int>& feed_mlvalue_idxs, const std:
   }
 
   // 3. handle the weights.
-  // We do this after the fetches to handle an edge case (possibly dubious) where a Constant is an output.
-  // The Constant gets lifted to an initializer so there's no Node producing the value as an output during Graph
-  // execution (i.e. Graph execution won't write the value to all_values_).
+  // We do this after the fetches to handle an edge case where an initializer is an output.
+  // e.g. A Constant node gets lifted to an initializer so there's no Node producing the value as an output during
+  // Graph execution (i.e. Graph execution won't write the value to all_values_).
   // A non-empty fetches vector will overwrite the actual weight in all_values_[ort_value_idx] if we did this earlier.
   // This makes the ONNX Constant test (onnx\backend\test\data\node\test_constant) happy as that
   // involves a graph with a single Constant node.
   for (const auto& entry : initializers) {
     int ort_value_index = entry.first;
-    all_values_[ort_value_index] = entry.second;
+
+    // if the initializer is an output we need to allocate or use a provided fetch buffer and copy the data
+    // so it can be returned to the caller.
+    if (IsOutput(ort_value_index)) {
+      const Tensor& src = entry.second.Get<Tensor>();  // all initializers in ONNX are tensors
+      OrtValue& dest = all_values_[ort_value_index];
+
+      if (!dest.IsAllocated()) {
+        // NOTE: This doesn't need to support ExecutionFrame custom allocators as they only come into play
+        // for a subgraph with an output of unknown shape that needs to be accumulated by the control flow node.
+        // If the initializer is providing the output, the shape is known.
+        AllocatorPtr allocator = GetAllocator(src.Location());
+
+        auto p_tensor = onnxruntime::make_unique<Tensor>(src.DataType(), src.Shape(), allocator);
+        auto ml_tensor = DataTypeImpl::GetType<Tensor>();
+        dest.Init(p_tensor.release(), ml_tensor, ml_tensor->GetDeleteFunc());
+      }
+
+      ORT_THROW_IF_ERROR(CopyTensor(src, *dest.GetMutable<Tensor>()));
```

Comment:
Curious how this plays out when initializer is on GPU? Will this copy be blocking at this point?

Reply:
It's a copy to the same device the initializer is on so it shouldn't block should it? The standard logic in utils::ExecuteGraph will handle any copies across devices if those are needed.